flanker/addresslib/address.py

```diff
@@ -73,27 +81,92 @@ def parse(address, addr_spec_only=False, metrics=False):
         None
     """
     mtimes = {'parsing': 0}
-
-    parser = flanker.addresslib.parser._AddressParser(False)
+    lexer = flanker.addresslib.lexer.lexer.clone()
+    if addr_spec_only:
+        parser = flanker.addresslib.parser.addr_spec_parser
+    else:
+        parser = flanker.addresslib.parser.mailbox_or_url_parser
+
+    # normalize inputs to bytestrings
+    if isinstance(address, unicode):
+        address = address.encode('utf-8')
+
+    # sanity checks
+    if not address:
+        return None, mtimes
+    if len(address) > MAX_ADDRESS_LENGTH:
+        log.warning('address exceeds maximum length of %s', MAX_ADDRESS_LENGTH)
+        return None, mtimes
 
     try:
-        # addr-spec only
-        if addr_spec_only:
-            bstart = time.time()
-            retval = parser.address_spec(address)
-            mtimes['parsing'] = time.time() - bstart
-            return retval, mtimes
-
-        # full address
         bstart = time.time()
-        retval = parser.address(address)
+        retval = _lift_parser_result(parser.parse(address, lexer=lexer))
         mtimes['parsing'] = time.time() - bstart
-        return retval, mtimes
+    except ply.lex.LexError as e:
+        log.warning(u'error in lexing: %s', e)
+        return None, mtimes
+    except ply.yacc.YaccError as e:
+        log.warning(u'error in parsing: %s', e)
+        return None, mtimes
+    except SyntaxError as e:
+        log.warning(u'error in parsing: %s', e)
+        return None, mtimes
+
+    return retval, mtimes
+
+
+@metrics_wrapper()
+def parse_discrete_list(address_list, metrics=False):
+    """
+    Given an string, returns an AddressList object (an iterable list
+    representing parsed email addresses and urls).
 
-    # supress any exceptions and return None
-    except flanker.addresslib.parser.ParserException:
+    Returns an AddressList object and optionally metrics on processing
+    time if requested.
+
+    Examples:
+        >>> address.parse_list('A <a@b>')
+        [A <a@b>]
+
+        >>> address.parse_list('A <a@b>, C <d@e>')
+        [A <a@b>, C <d@e>]
+
+        >>> address.parse_list('A <a@b>, C, D <d@e>')
+        None
+
+        >>> address.parse_list('A <a@b>, D <d@e>, http://localhost')
+        [A <a@b>, D <d@e>, http://localhost]
+    """
+    mtimes = {'parsing': 0}
+    lexer = flanker.addresslib.lexer.lexer.clone()
```

Comment:
It's `lexer.lexer` because its the `lexer` global variable in `lexer.py`. I might be able to clean that up by making the lexer/parser into a separate module. The problem with that is that PLY uses a lot of meta-programming to build the lexer and parsing tables so it's not as easy as copy-pasting the code into new files and I didn't think it was worth the effort. I think the import cleanup I did in response to another comment made the lexer usage more readable anyhow.